### PR TITLE
Enhance xcattest to support label search

### PR DESCRIPTION
**Pull request by hu-weihua**
Enhance ``xcattest`` to support label search.

The design document is https://github.com/xcat2/xcat-core/wiki/Test-Design-of-TestCase-Label

**Note**, Due to use ``perltidy`` to reformat script, so some code change just are about the format change, please ignore. 

UT:

```
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -h
Usage:
......
To list the information about label
    xcattest -l labelinfo
......
To list cases information that satisfy specific filter expression
    xcattest -l {caselist|caseinfo|casenum} -s "filter_expression"
To list information about cases in specific bundles and satisfy specific filter expression at same time
    xcattest -l {caselist|caseinfo|casenum} -b <bundle_list> -s "filter_expression"
......
To run test cases that satisfy specific filter expression
    xcattest -s "filter_expression"
....
Options:
....
    -s : Filter case by label. The acceptable value looks like 'label1+label2-label3|label4|label5', 'label1+label2-label3' works as sub expression.
         '|' means union filter, it has higher priority, expression1|expression2 means the case either satisfies expression1 or satisfies expression2.
         '+' means the case need have specific label. low priority.
         '-' means the case does not have specific label. low priority.
```

**Note**, if a case has had key word ``hcp``,``os``,``arch`` already, the label system will switch them to a label automatically.  Take below case for example:
```
start:install_xCAT_on_rhels_sles
os:redhat,  sles
label:xcat_install
cmd:date
end
```
Even without key word ``label``, this case will have label ``os=redhat``and  ``os=sles`` automatically. This design is for compatible with existed cases. 

Below bogus cases are used to test new ``xcattest``, if need to verify new code in your own environment, rename ``/opt/xcat/share/xcat/tools/autotest/testcase`` directory and copy below bogus cases to ``/opt/xcat/share/xcat/tools/autotest/testcase/boguscase``.
```
start:install_xCAT_on_rhels_sles
os:redhat, sles
label:xcat_install
cmd:date
end

start:install_xCAT_on_ubuntu
os:ubuntu
label:xcat_install
cmd:date
end

start:SN_setup_case
label:sn_diskful,provision
cmd:date
end

start:SN_diskless_setup_case
label: sn_diskless,provision
cmd:date
end

start:xcatd_start
label:mn_only, xcatd
cmd:date
end

start:xcatd_stop
label:mn_only,xcatd
cmd:date
end

start:lsdef_n
label:mn_only,db
cmd:date
end

start:mkdef_n
label:mn_only,db
cmd:date
end

start:reg_linux_diskfull_installation_flat
label:flat_cn_diskful,provision
os:linux
cmd:date
end

start:reg_linux_diskless_installation_flat
label:flat_cn_diskless,provision
os:linux
cmd:date
end

start:reg_linux_diskfull_installation_hierarchy
label:hierarchy_cn_diskful,provision
cmd:date
end

start:reg_linux_diskless_installation_hierarchy
label:hierarchy_cn_diskless, provision
cmd:date
end


start:xdcp_src_dst
label:cn_os_ready,dhcp
cmd:date
end

start:makedns_environment_check_forworder_mode
label:cn_os_ready,dns
cmd:date
end

start:rpower_status
hcp:openbmc, impi
label:cn_bmc_ready,hctrl_gen
cmd:date
end

start:rpower_suspend_OpenpowerBmc
hcp:openbmc
label:cn_bmc_ready,hctrl_openbmc
cmd:date
end

start:rpower_onstandby
arch:ppc64
label:cn_bmc_ready,hctrl_fsp
cmd:date
end

start:ib_invoke_provison
label:others,ib,invoke_provison
cmd:date
end

start:ib
label:others,ib
cmd:date
end

start:case_without_label
cmd:date
end
```

The test results are:
```
#For list all label information
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -l labelinfo
-------------------------------
The labels of cases:
-------------------------------
SN_diskless_setup_case                       sn_diskless,provision,
SN_setup_case                                sn_diskful,provision,
ib                                           others,ib,
ib_invoke_provison                           others,ib,invoke_provison,
install_xCAT_on_rhels_sles                   os=redhat,os=sles,xcat_install,
install_xCAT_on_ubuntu                       os=ubuntu,xcat_install,
lsdef_n                                      mn_only,db,
makedns_environment_check_forworder_mode     cn_os_ready,dns,
mkdef_n                                      mn_only,db,
reg_linux_diskfull_installation_flat         flat_cn_diskful,provision,os=linux,
reg_linux_diskfull_installation_hierarchy    hierarchy_cn_diskful,provision,
reg_linux_diskless_installation_flat         flat_cn_diskless,provision,os=linux,
reg_linux_diskless_installation_hierarchy    hierarchy_cn_diskless,provision,
rpower_onstandby                             arch=ppc64,cn_bmc_ready,hctrl_fsp,
rpower_status                                hcp=openbmc,hcp=impi,cn_bmc_ready,hctrl_gen,
rpower_suspend_OpenpowerBmc                  hcp=openbmc,cn_bmc_ready,hctrl_openbmc,
xcatd_start                                  mn_only,xcatd,
xcatd_stop                                   mn_only,xcatd,
xdcp_src_dst                                 cn_os_ready,dhcp,

-------------------------------
There are 28 different labels
The number of cases of each label:
-------------------------------
arch=ppc64               1
cn_bmc_ready             3
cn_os_ready              2
db                       2
dhcp                     1
dns                      1
flat_cn_diskful          1
flat_cn_diskless         1
hcp=impi                 1
hcp=openbmc              2
hctrl_fsp                1
hctrl_gen                1
hctrl_openbmc            1
hierarchy_cn_diskful     1
hierarchy_cn_diskless    1
ib                       2
invoke_provison          1
mn_only                  4
os=linux                 2
os=redhat                1
os=sles                  1
os=ubuntu                1
others                   2
provision                6
sn_diskful               1
sn_diskless              1
xcat_install             2
xcatd                    2
-------------------------------
There are 20 cases totaly, there are 19 cases with label.

#For generating a daily regression bundle
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -s "xcat_install+os=ubuntu|xcatd|flat_cn_diskful|cn_os_ready|ib-invoke_provison" -l caselist
install_xCAT_on_ubuntu
xcatd_start
xcatd_stop
reg_linux_diskfull_installation_flat
xdcp_src_dst
makedns_environment_check_forworder_mode
ib

#For generating a smoking test bundle
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -s "mn_only" -l caselist
xcatd_start
lsdef_n
xcatd_stop
mkdef_n

#Filter out all provision test cases 
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -s "provision" -l caselist
SN_setup_case
SN_diskless_setup_case
reg_linux_diskfull_installation_flat
reg_linux_diskless_installation_flat
reg_linux_diskfull_installation_hierarchy
reg_linux_diskless_installation_hierarchy

#Filter out all  test cases need to install CN/SN
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -s "provision|invoke_provison" -l caselist
SN_setup_case
SN_diskless_setup_case
reg_linux_diskfull_installation_flat
reg_linux_diskless_installation_flat
reg_linux_diskfull_installation_hierarchy
reg_linux_diskless_installation_hierarchy
ib_invoke_provison

#Filter out all cases against one component, take bogus IB for example
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -s "ib" -l caselist
ib
ib_invoke_provison

#Filter out all cases against one component but without provision CN
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -s "ib-invoke_provison" -l caselist
ib
```
All the test case filtered out are ordered by prerequisite order if they do not belong to the same level.

To run cases that satisfy specific filter expression
```
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -s "mn_only"
xCAT automated test started at Wed Jun 13 03:51:37 2018
******************************
To detect current test environment
******************************
Detecting: OS = linux
Warning: No compute node defined, can't get ARCH of compute node
Warning: No compute node defined, can't get HCP TYPE of compute node
******************************
loading test cases
******************************
To run:
xcatd_start
lsdef_n
xcatd_stop
mkdef_n
******************************
Start to run test cases
******************************
------START::xcatd_start::Time:Wed Jun 13 03:51:37 2018------

RUN:date [Wed Jun 13 03:51:37 2018]
ElapsedTime:0 sec
RETURN rc = 0
OUTPUT:
Wed Jun 13 03:51:37 EDT 2018

------END::xcatd_start::Passed::Time:Wed Jun 13 03:51:37 2018 ::Duration::0 sec------
------START::lsdef_n::Time:Wed Jun 13 03:51:37 2018------

RUN:date [Wed Jun 13 03:51:37 2018]
ElapsedTime:0 sec
RETURN rc = 0
OUTPUT:
Wed Jun 13 03:51:37 EDT 2018

------END::lsdef_n::Passed::Time:Wed Jun 13 03:51:37 2018 ::Duration::0 sec------
------START::xcatd_stop::Time:Wed Jun 13 03:51:37 2018------

RUN:date [Wed Jun 13 03:51:37 2018]
ElapsedTime:0 sec
RETURN rc = 0
OUTPUT:
Wed Jun 13 03:51:37 EDT 2018

------END::xcatd_stop::Passed::Time:Wed Jun 13 03:51:37 2018 ::Duration::0 sec------
------START::mkdef_n::Time:Wed Jun 13 03:51:37 2018------

RUN:date [Wed Jun 13 03:51:37 2018]
ElapsedTime:0 sec
RETURN rc = 0
OUTPUT:
Wed Jun 13 03:51:37 EDT 2018

------END::mkdef_n::Passed::Time:Wed Jun 13 03:51:37 2018 ::Duration::0 sec------
------Total: 4 , Failed: 0------

xCAT automated test finished at Wed Jun 13 03:51:37 2018
Please check results in the /home/xcattest_investigation//result/,
and see /home/xcattest_investigation//result//failedcases.20180613035137 file for failed cases.
see /home/xcattest_investigation//result//performance.report.20180613035137 file for time consumption
```

For error handling:
```
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -s errorlabel
There is no case to run, Please indicate case by option -b,-c,-t,-s
```

Complex filter
```
[root@c910f03c17k25 xcattest_investigation]# ./xcattest -s "mn_only"  -t mkdef_n,lsdef_n,ib  -l caselist
lsdef_n
mkdef_n
```

### CODE REVIEW

**Comment by whowutwut** [comment]
@hu-weihua FYI, we could avoid the `perltidy` changes mixing with the functional changes by using multiple commits.  Commit the code function in one commit, then at the end (or in the beginning) run `perltdy` and commit the changes in another commit, this way the review is easier to do..  IMHO

**Comment by whowutwut** [comment]
@hu-weihua I pulled down your  changes in this PR and built  it and installed onto a test machine, cannot successfully run it: 
 
```
[root@fs2vm112 ~]# xcattest -s "mn_only" 2>&1 | more
Use of uninitialized value $c in numeric gt (>) at /opt/xcat/bin/xcattest line 524.
Use of uninitialized value $c in numeric lt (<) at /opt/xcat/bin/xcattest line 527.
Use of uninitialized value $c in numeric gt (>) at /opt/xcat/bin/xcattest line 524.
Use of uninitialized value $c in numeric lt (<) at /opt/xcat/bin/xcattest line 527.
Use of uninitialized value $c in numeric gt (>) at /opt/xcat/bin/xcattest line 524.
Use of uninitialized value $c in numeric lt (<) at /opt/xcat/bin/xcattest line 527.
Use of uninitialized value $c in numeric gt (>) at /opt/xcat/bin/xcattest line 524.
Use of uninitialized value $c in numeric lt (<) at /opt/xcat/bin/xcattest line 527.
Use of uninitialized value $c in numeric gt (>) at /opt/xcat/bin/xcattest line 524.
Use of uninitialized value $c in numeric lt (<) at /opt/xcat/bin/xcattest line 527.
Use of uninitialized value $c in numeric gt (>) at /opt/xcat/bin/xcattest line 524.
Use of uninitialized value $c in numeric lt (<) at /opt/xcat/bin/xcattest line 527.
Use of uninitialized value $c in numeric gt (>) at /opt/xcat/bin/xcattest line 524.
Use of uninitialized value $c in numeric lt (<) at /opt/xcat/bin/xcattest line 527.
...
...
```

File: xCAT-test/xcattest [hunk 1]
```diff
@@ -52,19 +53,29 @@ my $RUN   = 1;
...
+my $caselist          = undef;
+my $cmdlist           = undef;
+my $list              = undef;
+my $restore           = 0;
+my $quiet             = 0;
+my $search_expression = undef;
+
+my %label_map;
+my $label_value = 1;
+my @label_order = (["xcat_install"],
```
Comment (robin2008) [inline xCAT-test/xcattest]: why using dict here,  `1 => ["xcat_install"]`, it looks more clear.

File: xCAT-test/xcattest [hunk 2]
```diff
@@ -1807,36 +1919,117 @@ sub print_table {
...
+    foreach (@$array_ref) {
+        push @tmp_arr, $_ unless ($_ eq $item);
     }
     @$array_ref = @tmp_arr;
 }
+
+
+sub filter_case_by_label{
+      my $error_ref=shift;
+        my $rst = load_case(\@cases_to_be_run, \@cases, \%case_name_index_map, $error_ref, $NORUN);
```
Comment (robin2008) [inline xCAT-test/xcattest]: why we design this method to include `load_case`,  I suggest you make it more general,  for example,  you can use `filter_case_by_label` after a load_case.

File: xCAT-test/xcattest [hunk 3]
```diff
@@ -1807,36 +1919,117 @@ sub print_table {
...
+        #        print "-----------------------------------\n";
+        #        print Dumper \@cases;
+        #        print "-----------------------------------\n";
+        #        print Dumper \%label_map;
+        #        print "===>exp = $search_expression\n";
+
+        my @filters = ();
+        my $index   = 0;
+        my @undefined_labels;
+        my @exps    = split('\|', $search_expression);
```
Comment (robin2008) [inline xCAT-test/xcattest]: Suggest you have a method `parse_filters` to get the filter rules,  then we can easily to UT on the function, and those codes could be shared to others if possible.

File: xCAT-test/xcattest [hunk 4]
```diff
@@ -1807,36 +1939,174 @@ sub print_table {
...
+    }
+
+    if (@undefined_labels) {
+        $$error_ref = "Label \"" . join(",", @undefined_labels) . "\" are not exist. Existed labels are: " . join(" ", keys(%label_map));
+        return 1;
+    }
+    return 0;
+}
+
+sub scan_existed_labels {
```
Comment (robin2008) [inline xCAT-test/xcattest]: What's purpose of this method?  could you give an function header here?  Looks like get all existing labels on case files.

File: xCAT-test/xcattest [hunk 5]
```diff
@@ -1807,36 +1939,174 @@ sub print_table {
...
+                }
+            } elsif ($line =~ /^label\s*:\s*(.+)/) {
+                push @labels, split(",", $1);
+            }
+
+            foreach my $t (@labels) {
+                unless (exists $label_map_ref->{$t}) {
+                    $label_map_ref->{$t} = $label_value;
+                    $label_value *= 2;
+                }
```
Comment (robin2008) [inline xCAT-test/xcattest]: Need a testing for max labels supported,  for example > 64 or 128 labels.

File: xCAT-test/xcattest [hunk 6]
```diff
@@ -1807,36 +1939,174 @@ sub print_table {
...
+    my $label_value = 1;
+    my @files       = ();
+    my @cmdfiles    = ();
+    get_files_recursive("$casedir", \@files);
+
+    my $fd = undef;
+    foreach my $file (@files) {
+        if (!open($fd, "<$file")) {
+            $$error_ref = "can't open $file:$!";
+            return 1;
```
Comment (robin2008) [inline xCAT-test/xcattest]: Is it the existing behavior, hit one error and failed for all?

File: xCAT-test/xcattest [hunk 7]
```diff
@@ -1807,36 +1939,174 @@ sub print_table {
...
+}
+
+sub scan_existed_labels {
+    my $label_map_ref = shift;
+    my $error_ref     = shift;
+
+    my $label_value = 1;
+    my @files       = ();
+    my @cmdfiles    = ();
+    get_files_recursive("$casedir", \@files);
```
Comment (robin2008) [inline xCAT-test/xcattest]: Is using `grep` better performance for getting all required match lines? and parse the output for each line.

File: xCAT-test/xcattest [hunk 8]
```diff
@@ -1807,36 +1939,174 @@ sub print_table {
...
+                    next;
+                }
+                push @{ $filters_ref->[$index] }, $label_map{$t};
+            }
+        }
+        $index += 1;
+    }
+
+    if (@undefined_labels) {
+        $$error_ref = "Label \"" . join(",", @undefined_labels) . "\" are not exist. Existed labels are: " . join(" ", keys(%label_map));
```
Comment (robin2008) [inline xCAT-test/xcattest]: Perhaps a warning is enough, to use a tag which does not exists is okay.

File: xCAT-test/xcattest [hunk 9]
```diff
@@ -136,20 +160,27 @@ if ($rst) {
     to_exit(1);
 }
 
+$rst = scan_existed_labels(\%label_map, \$error);
```
Comment (robin2008) [inline xCAT-test/xcattest]: If no `-s`, no need to run as user does not care about the tags.

File: xCAT-test/xcattest [hunk 10]
```diff
@@ -468,6 +540,25 @@ sub calculate_cases_to_be_run {
...
         }
     }
+
+
+    if($search_expression){
+        my @tmp_cases=();
+        my %tmp_case_name_index_map;
+        my @rest_cases_to_be_run=();
+
+        my $rst = load_case($cases_to_be_run_ref, \@tmp_cases, \%tmp_case_name_index_map, $error_ref, $NORUN);
```
Comment (robin2008) [inline xCAT-test/xcattest]: Function is okay, I am really wondering the performance,  as we can see it read all files three times when using tags.
1, scan all tags (only get tags, but no the relationship between case and tags)
2,load_case without run to do the filter (create the tag)
3,load rest cases to run

File: xCAT-test/xcattest (hunk 5 above)
Comment (hu-weihua) [reply xCAT-test/xcattest]: Refer to below code
```
my $base=1;
for(my $i=1; $i<=128; $i++){
   print "base[$i] = $base\n";
   $base*=2;
}
```
The output are:
```
base[1] = 1
base[2] = 2
......
base[63] = 4611686018427387904
base[64] = 9223372036854775808
base[65] = 1.84467440737096e+19
......
base[127] = 8.50705917302346e+37
base[128] = 1.70141183460469e+38
```
It seems we still can use the scientific counting method for calculation.

File: xCAT-test/xcattest [hunk 11]
```diff
@@ -1807,36 +1930,184 @@ sub print_table {
...
+    my $label_map_ref      = shift;
+    my $error_ref          = shift;
+
+    my $label_value = 1;
+    my @files       = ();
+    my @cmdfiles    = ();
+    get_files_recursive("$casedir", \@files);
+
+    foreach my $file (@files) {
+        my @output = runcmd("grep -E \"start:|hcp:|os:|arch:|label:\" $file");
```
Comment (robin2008) [inline xCAT-test/xcattest]: Maybe use the more strict one: `"^start:|^\s*hcp:|^\s*os:|^\s*arch:|^\s*label:"`

For multiple files grep, you can use 
```
grep -r -h -E "^start:|^\s*hcp:|^\s*os:|^\s*arch:|^\s*label:" ./*
```

**Comment by hu-weihua** [comment]
related task xcat2/xcat2-task-management#146 and xcat2/xcat2-task-management#170

**Pull request merged**